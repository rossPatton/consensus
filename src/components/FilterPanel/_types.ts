export type tProps = {
  className?: string,
  filterOptions?: {key: string, display: string}[],
  onCategoryChange?: (ev: React.ChangeEvent<HTMLSelectElement>) => void,
  onFilterOptionChange?: (ev: React.ChangeEvent<HTMLSelectElement>) => void,
  onPublishedFilterChange?: (ev: React.ChangeEvent<HTMLSelectElement>) => void,
  onRoleFilterChange?: (ev: React.ChangeEvent<HTMLSelectElement>) => void,
  onSearchChange: (event: React.ChangeEvent<HTMLInputElement>) => void,
  placeholder?: string,
};
