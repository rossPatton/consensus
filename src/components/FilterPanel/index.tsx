import cx from 'classnames';
import React, {memo} from 'react';

import {categories, roles} from '~app/constants';

import {tProps} from './_types';

/**
 * @description the component half of the search filter container. use together
 *
*/
const FilterPanel = memo((props: tProps) => (
  <div
    className={cx({
      [props.className]: !!props.className,
      'flex flex-col d:flex-row': true,
      'mb-2 d:mb-3': !props.className,
    })}>
    {props.filterOptions && (
      <select
        className="w-full d:w-3/12 mb-1 d:mb-0 d:mr-2"
        onBlur={props.onFilterOptionChange}
        onChange={props.onFilterOptionChange}>
        {props.filterOptions.map(opt => (
          <option key={opt.key} value={opt.key}>
            {opt.display}
          </option>
        ))}
      </select>
    )}
    {props.onCategoryChange && (
      <select
        className="w-full d:w-3/12 mb-1 d:mb-0 d:mr-2"
        onBlur={props.onCategoryChange}
        onChange={props.onCategoryChange}>
        <option value="">
          Show all
        </option>
        {categories.map(({display}) => (
          <option key={display} value={display}>
            {display}
          </option>
        ))}
      </select>
    )}
    {/*
      @TODO this is misleadingly named -> should be past/future filter
      published should mean isDraft or not
    */}
    {props.onPublishedFilterChange && (
      <select
        className="w-full d:w-3/12 mb-1 d:mb-0 d:mr-2"
        defaultValue={props.publishedFilter}
        onBlur={props.onPublishedFilterChange}
        onChange={props.onPublishedFilterChange}>
        <option value="upcoming">
          Filter meetings
        </option>
        {[
          {display: 'Show all', value: 'n/a'},
          {display: 'Upcoming', value: 'upcoming'},
          {display: 'Past', value: 'past'},
        ].map(({display, value}) => (
          <option key={display} value={value}>
            {display}
          </option>
        ))}
      </select>
    )}
    {props.onRoleFilterChange && (
      <select
        className="w-full d:w-3/12 mb-1 d:mb-0 d:mr-2"
        onBlur={props.onRoleFilterChange}
        onChange={props.onRoleFilterChange}>
        {['n/a', ...roles].map(role => {
          let display = 'Show all';
          if (role === 'member') {
            display = props.memberName;
          } else if (role === 'facilitator') {
            display = props.modName;
          }

          return (
            <option
              key={role}
              value={role}>
              {display}
            </option>
          );
        })}
      </select>
    )}
    <label
      className="w-full"
      htmlFor="searchFilter">
      <input
        spellCheck
        type="search"
        id={props.id || 'searchFilter'}
        className="w-full"
        onChange={props.onSearchChange}
        placeholder={props.placeholder || 'Filter all results'}
      />
    </label>
  </div>
));

export default FilterPanel;
