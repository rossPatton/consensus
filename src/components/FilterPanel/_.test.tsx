import React from 'react';
import render from 'react-test-renderer';

import FilterPanel from '.';

describe('components/FilterPanel', () => {
  it('renders', () => {
    const component = render.create((
      <FilterPanel
        onSearchChange={jest.fn()}
      />
    ));

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with standard props', () => {
    const component = render.create((
      <FilterPanel
        className="test"
        filterOptions={[{key: 'test', display: 'Display'}]}
        memberName="dude"
        modName="other dude"
        placeholder="filter options here"
        onSearchChange={jest.fn()}
      />
    ));

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with categories', () => {
    const component = render.create((
      <FilterPanel
        className="test"
        filterOptions={[{key: 'test', display: 'Display'}]}
        onCategoryChange={jest.fn()}
        memberName="dude"
        modName="other dude"
        placeholder="filter options here"
        onSearchChange={jest.fn()}
      />
    ));

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with published filter', () => {
    const component = render.create((
      <FilterPanel
        className="test"
        filterOptions={[{key: 'test', display: 'Display'}]}
        onPublishedFilterChange={jest.fn()}
        memberName="dude"
        modName="other dude"
        placeholder="filter options here"
        onSearchChange={jest.fn()}
      />
    ));

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with roles filter', () => {
    const component = render.create((
      <FilterPanel
        className="test"
        filterOptions={[{key: 'test', display: 'Display'}]}
        onRoleFilterChange={jest.fn()}
        memberName="dude"
        modName="other dude"
        placeholder="filter options here"
        onSearchChange={jest.fn()}
      />
    ));

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
