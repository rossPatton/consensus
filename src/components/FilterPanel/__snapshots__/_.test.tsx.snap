// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/FilterPanel renders 1`] = `
<div
  className="flex flex-col d:flex-row mb-2 d:mb-3"
>
  <label
    className="w-full"
    htmlFor="searchFilter"
  >
    <input
      className="w-full"
      id="searchFilter"
      onChange={[MockFunction]}
      placeholder="Filter all results"
      spellCheck={true}
      type="search"
    />
  </label>
</div>
`;

exports[`components/FilterPanel renders with categories 1`] = `
<div
  className="test flex flex-col d:flex-row"
>
  <select
    className="w-full d:w-3/12 mb-1 d:mb-0 d:mr-2"
  >
    <option
      value="test"
    >
      Display
    </option>
  </select>
  <select
    className="w-full d:w-3/12 mb-1 d:mb-0 d:mr-2"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
  >
    <option
      value=""
    >
      Show all
    </option>
    <option
      value="Community"
    >
      Community
    </option>
    <option
      value="Cooperative"
    >
      Cooperative
    </option>
    <option
      value="Unions and Labor"
    >
      Unions and Labor
    </option>
    <option
      value="Political"
    >
      Political
    </option>
  </select>
  <label
    className="w-full"
    htmlFor="searchFilter"
  >
    <input
      className="w-full"
      id="searchFilter"
      onChange={[MockFunction]}
      placeholder="filter options here"
      spellCheck={true}
      type="search"
    />
  </label>
</div>
`;

exports[`components/FilterPanel renders with published filter 1`] = `
<div
  className="test flex flex-col d:flex-row"
>
  <select
    className="w-full d:w-3/12 mb-1 d:mb-0 d:mr-2"
  >
    <option
      value="test"
    >
      Display
    </option>
  </select>
  <select
    className="w-full d:w-3/12 mb-1 d:mb-0 d:mr-2"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
  >
    <option
      value="n/a"
    >
      Filter meetings
    </option>
    <option
      value="n/a"
    >
      Show all meetings
    </option>
    <option
      value="upcoming"
    >
      Upcoming
    </option>
    <option
      value="past"
    >
      Past
    </option>
  </select>
  <label
    className="w-full"
    htmlFor="searchFilter"
  >
    <input
      className="w-full"
      id="searchFilter"
      onChange={[MockFunction]}
      placeholder="filter options here"
      spellCheck={true}
      type="search"
    />
  </label>
</div>
`;

exports[`components/FilterPanel renders with roles filter 1`] = `
<div
  className="test flex flex-col d:flex-row"
>
  <select
    className="w-full d:w-3/12 mb-1 d:mb-0 d:mr-2"
  >
    <option
      value="test"
    >
      Display
    </option>
  </select>
  <select
    className="w-full d:w-3/12 mb-1 d:mb-0 d:mr-2"
    onBlur={[MockFunction]}
    onChange={[MockFunction]}
  >
    <option
      value="n/a"
    >
      Show all
    </option>
    <option
      value="member"
    >
      dude
    </option>
    <option
      value="facilitator"
    >
      other dude
    </option>
  </select>
  <label
    className="w-full"
    htmlFor="searchFilter"
  >
    <input
      className="w-full"
      id="searchFilter"
      onChange={[MockFunction]}
      placeholder="filter options here"
      spellCheck={true}
      type="search"
    />
  </label>
</div>
`;

exports[`components/FilterPanel renders with standard props 1`] = `
<div
  className="test flex flex-col d:flex-row"
>
  <select
    className="w-full d:w-3/12 mb-1 d:mb-0 d:mr-2"
  >
    <option
      value="test"
    >
      Display
    </option>
  </select>
  <label
    className="w-full"
    htmlFor="searchFilter"
  >
    <input
      className="w-full"
      id="searchFilter"
      onChange={[MockFunction]}
      placeholder="filter options here"
      spellCheck={true}
      type="search"
    />
  </label>
</div>
`;
